pkg/agent/memberlist/cluster_test.go

```diff
@@ -0,0 +1,626 @@
+						hitCount++
+						seqSum += i
+					}
+				}
+			}
+			assert.Equal(t, expectEgressSeqSum, seqSum, "Egress seq")
+			for _, node := range nodes {
+				t.Logf("Node (%s) Egress: %#v", node, nodeEgress[node])
+			}
+			assert.Equal(t, egressNum, hitCount, "Egress total num should be 30")
+		})
+	}
+}
+
+func TestCluster_ShouldSelectEgress(t *testing.T) {
+	testCases := []struct {
+		name         string
+		nodeNum      int
+		egressIP     string
+		expectedNode string
+	}{
+		{
+			name:         "Select Node from 0 nodes",
+			nodeNum:      0,
+			egressIP:     "1.1.1.1",
+			expectedNode: "",
+		},
+		{
+			name:         "Select Node from 1 nodes",
+			nodeNum:      1,
+			egressIP:     "1.1.1.1",
+			expectedNode: "node-0",
+		},
+		{
+			name:         "Select Node from 3 nodes",
+			nodeNum:      3,
+			egressIP:     "1.1.1.1",
+			expectedNode: "node-1",
+		},
+		{
+			name:         "Select Node from 10 nodes",
+			nodeNum:      10,
+			egressIP:     "1.1.1.1",
+			expectedNode: "node-1",
+		},
+		{
+			name:         "Select Node from 100 nodes",
+			nodeNum:      100,
+			egressIP:     "1.1.1.1",
+			expectedNode: "node-79",
+		},
+	}
+	for _, tCase := range testCases {
+		t.Run(tCase.name, func(t *testing.T) {
+			genLocalNodeCluster := func(nodeNme, eipName string, nodes []string) *Cluster {
+				cluster := &Cluster{
+					mList:             &memberlist.Memberlist{},
+					nodeName:          nodeNme,
+					consistentHashMap: make(map[string]*consistenthash.Map),
+				}
+				cluster.consistentHashMap[eipName] = consistenthash.New(defaultVirtualNodeReplicas, nil)
+				cluster.consistentHashMap[eipName].Add(nodes...)
+				return cluster
+			}
+
+			genNodes := func(n int) []string {
+				nodes := make([]string, n)
+				for i := 0; i < n; i++ {
+					nodes[i] = fmt.Sprintf("node-%d", i)
+				}
+				return nodes
+			}
+
+			fakeEIPName := "fakeExternalIPPool"
+			fakeEgress := &crdv1a2.Egress{
+				ObjectMeta: metav1.ObjectMeta{Name: "fakeEgress"},
+				Spec:       crdv1a2.EgressSpec{ExternalIPPool: fakeEIPName, EgressIP: tCase.egressIP},
+			}
+			nodes := genNodes(tCase.nodeNum)
+			fakeCluster := genLocalNodeCluster(tCase.expectedNode, fakeEIPName, nodes)
+			selected, err := fakeCluster.ShouldSelectEgress(fakeEgress)
+			assert.NoError(t, err)
+			assert.Equal(t, true, selected, "Selected Node for Egress not match")
+		})
+	}
+}
+
+// BenchmarkCluster_ShouldSelect
+// BenchmarkCluster_ShouldSelect/select_node_from_1000_alive_nodes
+// BenchmarkCluster_ShouldSelect/select_node_from_1000_alive_nodes-16         	12263878	        95.5 ns/op
+// BenchmarkCluster_ShouldSelect/select_node_from_100_alive_nodes
+// BenchmarkCluster_ShouldSelect/select_node_from_100_alive_nodes-16          	13036746	       103 ns/op
+// BenchmarkCluster_ShouldSelect/select_node_from_10_alive_nodes
+// BenchmarkCluster_ShouldSelect/select_node_from_10_alive_nodes-16           	14923483	        77.5 ns/op
+func BenchmarkCluster_ShouldSelect(b *testing.B) {
```

Comment:
The benchmark doesn't really measure a method in source code, it calls one step in `ShouldSelectEgress` but also counts the time spent on creating a consistenthashMap. Can it focus one method i.e. ShouldSelectEgress? I don't think we need many input matrix either (as long as it's efficient in 1000 nodes, no one will worry about 10 nodes).

Reply:
done